src/lib.rs

```diff
@@ -953,6 +958,29 @@ pub struct Hasher {
     cv_stack: ArrayVec<CVBytes, { MAX_DEPTH + 1 }>,
 }
 
+#[cfg(feature = "zeroize")]
+impl zeroize::Zeroize for Hasher {
+    fn zeroize(&mut self) {
+        use std::mem::MaybeUninit;
+        self.key.zeroize();
+        self.chunk_state.zeroize();
+        // Set the Vec's length to 0
+        self.cv_stack.clear();
+
+        // Zero the full capacity of `ArrayVec`.
+        // Safety:
+        // This is safe because MaybeUninit ignores all invariants,
+        // so we can create a slice of MaybeUninit<CVBytes> using the full capacity of the Vec
+        let uninit_cv_stack = unsafe {
+            core::slice::from_raw_parts_mut(
+                self.cv_stack.as_mut_ptr().cast::<MaybeUninit<CVBytes>>(),
+                self.cv_stack.capacity(),
+            )
+        };
+        uninit_cv_stack.zeroize();
```

Comment:
I wonder if either `Zeroize` or `ArrayVec` would consider supporting the other directly. They're both pretty popular crates, and this seems like the sort of sharp edge that a lot of folks will run into.

In any case, I think we could do this safely by just appending a bunch of all-zero CVs to the array until `.is_full()`, and then zeroizing a mutable slice over the whole `ArrayVec`. Annoyingly, that doesn't seem to work directly on the slice type, because `Zeroize` over slices depends on `DefaultIsZeroes`, and arrays of ints don't implement `DefaultIsZeroes` (no idea why). But we can work around that by casting the slice into an array reference, like this:

```rust
while !self.cv_stack.is_full() {
    self.cv_stack.push([0; 32]);
}
let cv_stack_array: &mut [CVBytes; MAX_DEPTH + 1] =
    (&mut self.cv_stack[..]).try_into().unwrap();
cv_stack_array.zeroize();
```

What do you think?

Whatever we decide to do here, we should definitely add a test to exercise it :)

Reply:
1. I can make a PR to `ArrayVec` to see if they'll land it.
2. General arrays don't implement `DefaultIsZeroes` because they don't implement `Default` (They implement up to `[T; 32]` ).
3. I like your no unsafe solution, I did not know you can `try_into()` a reference to a slice into a reference to an array (I thought only owned values are possible) :)

Reply:
Ooo it looks like number 1 might work out, that's awesome.